src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+            ret.element(STROKE, stroke.getColor().evaluate(feature));
+            ret.element(STROKE_WIDTH, stroke.getWidth().evaluate(feature));
+        }
+        if (stroke.getOpacity() != null) {
+            ret.element(STROKE_OPACITY, printExpression(stroke.getOpacity(), Number.class));
+        }
+        if (stroke.getLineCap() != null) {
+            ret.element(STROKE_LINECAP, printExpression(stroke.getLineCap(), String.class));
+        }
+        if (stroke.getLineJoin() != null) {
+            ret.element(STROKE_LINEJOIN, printExpression(stroke.getLineJoin(), String.class));
+        }
+        List<Expression> dashArray = stroke.dashArray();
+        if (dashArray != null && !dashArray.isEmpty()) {
+            JSONArray dArray = new JSONArray();
+            for (Expression e : dashArray) {
+                dArray.add(e.evaluate(feature));
+            }
+            ret.element(STROKE_DASHARRAY, dArray);
+            Expression dashOffset = stroke.getDashOffset();
+            if (dashOffset != null) {
+                ret.element(STROKE_DASHOFFSET, dashOffset.evaluate(feature));
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * @param symbolizer
+     * @return
+     */
+    private JSONObject processSymbolizer(Symbolizer symbolizer) {
+        JSONObject ret = new JSONObject();
+        String name = symbolizer.getName();
+        if (name != null && !name.isEmpty()) {
+            ret.element(NAME, name);
+        }
+        Unit<Length> uom = symbolizer.getUnitOfMeasure();
+        if (uom != null) {
+            ret.element(UOM, uom.toString());
+        }
+        Description desc = symbolizer.getDescription();
+        if (desc != null) {
+            InternationalString title = desc.getTitle();
+            if (title != null) {
+                ret.element(TITLE, title.toString());
+            }
+            InternationalString abs = desc.getAbstract();
+            if (abs != null) {
+                ret.element(ABSTRACT, abs.toString());
+            }
+        }
+        Expression geometry = symbolizer.getGeometry();
+        if (geometry != null) {
+            ret.element(GEOMETRY, printExpression(geometry, Geometry.class));
+        }
+        if (symbolizer instanceof PointSymbolizer) {
+            ret = processPointSymbolizer(ret, (PointSymbolizer) symbolizer);
+        } else if (symbolizer instanceof LineSymbolizer) {
+            ret = processLineSymbolizer(ret, (LineSymbolizer) symbolizer);
+        } else if (symbolizer instanceof PolygonSymbolizer) {
+            ret = processPolygonSymbolizer(ret, (PolygonSymbolizer) symbolizer);
+        } else if (symbolizer instanceof RasterSymbolizer) {
+            ret = processRasterSymbolizer(ret, (RasterSymbolizer) symbolizer);
+        } else if (symbolizer instanceof TextSymbolizer) {
+            ret = processTextSymbolizer(ret, (TextSymbolizer) symbolizer);
+        }
```

Comment:
I'd add a paranoid check for unkonwn symbolizer types, either logging or throwing an exception.